internal/failure/failure.go

```diff
@@ -0,0 +1,268 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package failure
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"text/scanner"
+)
+
+type (
+	// Field references a Failure field.
+	Field int
+
+	// ID references a Failure identifier.
+	ID int
+)
+
+const (
+	// Filename references the Filename field of a Failure.
+	Filename Field = iota
+	// Line references the Line field of a Failure.
+	Line
+	// Column references the Column field of a Failure.
+	Column
+	// Identifier references the ID field of a Failure.
+	Identifier
+	// Message references the Message field of a Failure.
+	Message
+
+	// Error identifies a generic Failure.
+	Error ID = iota
+	// Format identifies a diff format Failure.
+	Format
+	// Lint identifies a lint Failure.
+	Lint
+	// Proto identitifes an invalid proto Failure.
+	Proto
+
+	// DefaultFilename is the default Failure filename.
+	DefaultFilename = "<input>"
+	// DefaultLine is the default Failure line.
+	DefaultLine = 1
+	// DefaultColumn is the default Failure column number.
+	DefaultColumn = 1
+)
+
+var (
+	_defaultFields = []Field{
+		Filename,
+		Line,
+		Column,
+		Message,
+	}
+	_stringToField = map[string]Field{
+		"filename": Filename,
+		"line":     Line,
+		"column":   Column,
+		"id":       Identifier,
+		"message":  Message,
+	}
+	_IDToString = map[ID]string{
+		Error:  "ERROR",
+		Format: "FORMAT",
+		Lint:   "LINT",
+		Proto:  "PROTO",
+	}
+)
+
+// String implements fmt.Stringer.
+func (id ID) String() string {
+	if s, ok := _IDToString[id]; ok {
+		return s
+	}
+	return strconv.Itoa(int(id))
+}
+
+// ParseFields parses Fields from the given string.
+// Fields are expected to be colon-separated in the given string.
+// Input is case-insensitive. If the string is empty, _defaultFields
+// will be returned.
+func ParseFields(s string) ([]Field, error) {
+	if len(s) == 0 {
+		return _defaultFields, nil
+	}
+	fields := strings.Split(s, ":")
+	failureFields := make([]Field, len(fields))
+	for i, f := range fields {
+		ff, err := parseField(f)
+		if err != nil {
+			return nil, err
+		}
+		failureFields[i] = ff
+	}
+	return failureFields, nil
+}
+
+// Failure is a failure with a position in text.
+type Failure struct {
```

Comment:
This will fail lint as Golang says foo.Foo is not allowed. For this and other reasons ill elaborate on tomorrow, I think this should be reverted to text. Failure is a noun, which doesn’t work well for packages.